ospfd/ospf_asbr.c

```diff
@@ -305,3 +307,764 @@ void ospf_redistribute_withdraw(struct ospf *ospf, uint8_t type,
+			__func__, inet_ntoa(ei->p.prefix), ei->p.prefixlen,
+			inet_ntoa(aggr->p.prefix), aggr->p.prefixlen);
+	hash_get(aggr->match_extnl_hash, ei, hash_alloc_intern);
+	ei->aggr_route = aggr;
+}
+
+struct ospf_lsa *ospf_originate_summary_lsa(struct ospf *ospf,
+					    struct ospf_external_aggr_rt *aggr,
+					    struct external_info *ei)
+{
+	struct ospf_lsa *lsa;
+	struct external_info ei_aggr;
+	struct as_external_lsa *asel;
+	struct ospf_external_aggr_rt *old_aggr;
+	route_tag_t tag = 0;
+
+	if (IS_DEBUG_OSPF(lsa, EXTNL_LSA_AGGR))
+		zlog_debug("%s: Prepare to originate Summary route(%s/%d)",
+			   __func__, inet_ntoa(aggr->p.prefix),
+			   aggr->p.prefixlen);
+
+	/* This case to handle when the overlapping aggregator address
+	 * is availbe.Best match will be considered.So need to delink
+	 * from old aggregator and link to the new aggr.
+	 */
+	if (ei->aggr_route) {
+		if (ei->aggr_route != aggr) {
+			old_aggr = ei->aggr_route;
+			ospf_unlink_ei_from_aggr(ospf, old_aggr, ei);
+		}
+	}
+
+	/* Add the external route to hash table */
+	ospf_link_ei_to_aggr(aggr, ei);
+
+	lsa = ospf_external_info_find_lsa(ospf, &aggr->p);
+	/* Dont originate external LSA,
+	 * If it is configured not to advertise.
+	 */
+	if (CHECK_FLAG(aggr->flags, OSPF_EXTERNAL_AGGRT_NO_ADVERTISE)) {
+		/* If it is already originated as external LSA,
+		 * But, it is configured not to advertise then
+		 * flush the originated external lsa.
+		 */
+		if (lsa)
+			ospf_external_lsa_flush(ospf, 0, &aggr->p, 0);
+		UNSET_FLAG(aggr->flags, OSPF_EXTERNAL_AGGRT_ORIGINATED);
+
+		if (IS_DEBUG_OSPF(lsa, EXTNL_LSA_AGGR))
+			zlog_debug(
+				"%s: Don't originate the summary address,It is configured to not-advertise.",
+				__func__);
+		return NULL;
+	}
+
+	/* Prepare the extrenal_info for aggregator */
+	memset(&ei_aggr, 0, sizeof(struct external_info));
+	ei_aggr.p = aggr->p;
+	ei_aggr.tag = aggr->tag;
+	ei_aggr.type = 0;
+	ei_aggr.instance = ospf->instance;
+	ei_aggr.route_map_set.metric = -1;
+	ei_aggr.route_map_set.metric_type = -1;
+
+	/* Summary route already originated,
+	 * So, Do nothing.
+	 */
+	if (CHECK_FLAG(aggr->flags, OSPF_EXTERNAL_AGGRT_ORIGINATED)) {
+		if (!lsa) {
+			flog_warn(EC_OSPF_LSA_MISSING,
+				  "%s: Could not refresh/originate %s/%d",
+				  __func__, inet_ntoa(aggr->p.prefix),
+				  aggr->p.prefixlen);
+			return NULL;
+		}
+
+		asel = (struct as_external_lsa *)lsa->data;
+		tag = (unsigned long)ntohl(asel->e[0].route_tag);
+
+		/* If tag modified , then re-originate the route
+		 * with modified tag details.
+		 */
+		if (tag != ei_aggr.tag) {
+			if (IS_DEBUG_OSPF(lsa, EXTNL_LSA_AGGR))
+				zlog_debug(
+					"%s: Route tag changed(old:%d new:%d,So refresh the summary route.(%s/%d)",
+					__func__, tag, ei_aggr.tag,
+					inet_ntoa(aggr->p.prefix),
+					aggr->p.prefixlen);
+
+			ospf_external_lsa_refresh(ospf, lsa, &ei_aggr,
+						  LSA_REFRESH_FORCE, 1);
+		}
+		return lsa;
+	}
+
+	if (lsa && IS_LSA_MAXAGE(lsa)) {
```

Comment:
LSA could be refresh before reaching MAX_AGE. Look at other part of the code where LSA are refreshed sooner e.g. when reaching MAX_AGE/2. In fact, there is a risk that a neighbor router remove the LSA which reach MAX_AGE before the advertised router refresh it. There is also a risk that the LSA is removed from the LSDB by the LSA garbage collector that remove all LSA that reach MAX_AGE it if runs before this function.

Reply:
Here, I am not refreshing the normal LSA. 
This is special case, Refreshing a LSA if it is in max age. Normally , If any LSA need to be flushed , it first will be added to MAX age queue with 3600 as age. After max-age delay time , it will be flushed from local db. Here I am trying to handle such LSAs. 
Lets take  the exact  scenario , An external route(addr 10.1.1.0/24) got deleted and subsequently the corresponding external LSA added to the max age queue.  At this time, If a summary route configured with same address ( 10.1.1.0/24), this will not be originated since the similar LSA present in the LSDB with max age . In order to send the summarised LSA to AS, this existing LSA need to be refreshed.

Reply:
OK. I understand. But, does it not safer to let the old LSA deleted and create a new LSA with the new summary route instead of refreshing the old one? I mean that the new summary route could differ from the old one. But, if this is not a realistic scenario, refreshing is the best to do as it is faster and avoid free/alloc dynamic memory.